src/printer.js

```diff
@@ -2019,71 +2022,170 @@ function printMemberChain(node, options, print) {
+      }
+    } else {
+      children.push(print(childPath));
+
+      // add a line unless it's followed by a JSX newline
+      let next = n.children[i + 1];
+      if (!(next && /^\s*\n/.test(next.value))) {
+        children.push(softline);
+      }
+    }
+  }, "children");
+
+  return children;
+}
+
 function printJSXElement(path, options, print) {
   const n = path.getValue();
-  var openingLines = path.call(print, "openingElement");
 
-  let elem;
+  // turn <div></div> into <div />
+  if (isEmptyJSXElement(n)) {
+    n.openingElement.selfClosing = true;
+    delete n.closingElement;
+  }
+
+  // if no children, just print the opening element
+  const openingLines = path.call(print, "openingElement");
   if (n.openingElement.selfClosing) {
     assert.ok(!n.closingElement);
+    return openingLines;
+  }
 
-    elem = openingLines;
-  } else {
-    var children = [];
+  const children = printJSXChildren(path, options, print);
+  let hasAnyHardLine = false;
 
-    path.map(
-      function(childPath) {
-        var child = childPath.getValue();
+  // trim trailing whitespace, recording if there was a hardline
+  while (children.length && util.getLast(children).type === "line") {
+    if (hasHardLine(util.getLast(children))) hasAnyHardLine = true;
+    children.pop();
+  }
 
-        if (
-          namedTypes.Literal.check(child) && typeof child.value === "string"
-        ) {
-          if (/\S/.test(child.value)) {
-            const beginBreak = child.value.match(/^\s*\n/);
-            const endBreak = child.value.match(/\n\s*$/);
-
-            children.push(
-              beginBreak ? hardline : "",
-              child.value.replace(/^\s+|\s+$/g, endBreak ? "" : " "),
-              endBreak ? hardline : ""
-            );
-          } else if (/\n/.test(child.value)) {
-            children.push(hardline);
-          }
-        } else {
-          children.push(print(childPath));
-        }
-      },
-      "children"
-    );
+  // trim leading whitespace, recording if there was a hardline
+  while (children.length && children[0].type === "line") {
+    if (hasHardLine(children[0])) hasAnyHardLine = true;
+    children.shift();
+  }
 
-    var mostChildren = children.slice(0, -1);
-    var closingLines = path.call(print, "closingElement");
+  //  prefill leading newline, since this is only used in multiLineElem
+  let childrenGroupedByLine = [ hardline ];
+  if (children.length === 1) {
+    // no need for fanciness if just one thing, and lone jsx-whitespace needs this
+    childrenGroupedByLine.push(children[0]);
 
-    const firstChild = children[(0)];
-    const lastChild = util.getLast(children);
-    const beginBreak = firstChild && firstChild.type === "line";
-    const endBreak = lastChild && lastChild.type === "line";
+    if (!hasAnyHardLine && hasHardLine(children[0])) hasAnyHardLine = true;
+  } else {
+    // group by line, recording if there was a hardline.
+    childrenGroupedByLine.push(group(concat([])));
```

Comment:
this initializes the first group... it's kind of ugly code, and used to be prettier until I handled the jsx-whitespace case above with this `if`